### ISSUE TITLE
Null pointer check inconsistency in ContextualCheckBlock

### ISSUE THREAD

**Issue by awemany** [gandrewstone/bitcoin#549]
In this snippet from main.cpp:

```

bool ContextualCheckBlock(const CBlock& block, CValidationState& state, CBlockIndex * const pindexPrev)
{
    const int nHeight = pindexPrev == NULL ? 0 : pindexPrev->nHeight + 1;
    const Consensus::Params& consensusParams = Params().GetConsensus();

    // Start enforcing BIP113 (Median Time Past) using versionbits logic.
    int nLockTimeFlags = 0;
    if (VersionBitsState(pindexPrev, consensusParams, Consensus::DEPLOYMENT_CSV, versionbitscache) == THRESHOLD_ACTIVE) {
        nLockTimeFlags |= LOCKTIME_MEDIAN_TIME_PAST;
    }

    int64_t nLockTimeCutoff = (nLockTimeFlags & LOCKTIME_MEDIAN_TIME_PAST)
                              ? pindexPrev->GetMedianTimePast()
                              : block.GetBlockTime();
```

You can see that pindexPrev is check for NULL pointer above. However, below, the code assumes the pointer to be non-null when VersionBitsState sets LOCKTIME_MEDIAN_TIME_PAST as it dereferences it.

This appears to be fragile code - and from further exploration, it is not entirely obvious to me that VersionBitsState(..) will never set that flag in case pindexPrev is NULL.

I do not feel familiar enough with the entangled main.cpp validation web to be sure to have the right answer on how to fix this.

**Comment by Greg-Griffith** [comment]
Someone correct me if im wrong.... But, i believe the following to be the correct answer to your question.

if a block with a pindexPrev of NULL is put through the VersionBitsState function,  you will return the state of THRESHOLD_DEFINED. This is guaranteed from this part of the while loop inside that function

    while (cache.count(pindexPrev) == 0) {
        if (pindexPrev == NULL) {
            // The genesis block is by definition defined.
            cache[pindexPrev] = THRESHOLD_DEFINED;
            break;
        }  

because with a pindexPrev of NULL you will never have a cache count that is not 0 from the start. and then it pretty much returns that state right after that. 

Due to C++'s default values for enums, the value for THRESHOLD_DEFINED is 0 and when evaluating for true or false on return, C++ defines anything non 0 to be true, and 0 to be false.
 
This will result in nLockTimeFlags still being 0 when "nLockTimeFlags & LOCKTIME_MEDIAN_TIME_PAST" is evaluated, again returning false. so block.GetBlockTime() will always be used instead of pindexPrev->GetMedianTimePast(). 
So there is no need to worry that VersionBitsState will ever set that flag when given a block with a pindexPrev of NULL.

That was kind of a rough explanation. hope it helped a little bit though.

**Comment by ftrader** [comment]
I agree with the explanation given by @Greg-Griffith here.

> Due to C++'s default values for enums, the value for THRESHOLD_DEFINED is 0 and when evaluating for true or false on return, C++ defines anything non 0 to be true, and 0 to be false.

Personally, I dislike it when code relies on such implicit behavior, and would prefer it to verbosely testing against THRESHOLD_DEFINED. It just makes it easier to understand what's going on.

**Comment by awemany** [comment]
@Greg-Griffith : Thank you for that detailed explanation. Yes I can follow it and agree it looks like the null pointer dereference is impossible.  It also depends on the state of the ThresholdConditionCache, but that also seems to work out.

Will add a PR now to make this explicit.

**Comment by awemany** [comment]
Solved by #845.
